Medusa not stopping cassandra as service properly.

While I am restoring a single node backup on same node, medusa is not stopping cassandra properly. I doubt it is removing "commitlogs" folder while cassandra shutdown (/etc/init.d/cassandra stop) is happening. So, shutdown is not clean. After restore, cassandra is not starting up as service (/etc/init.d/cassandra start). So I have to run "cassandra stop" and "cassandra start" again. I am on CentOS 6.10.

Medusa has to wait till cassandra is shutdown gracefully before removing commitlogs folder.

[root@localhost ~]# medusa restore-node --backup-name test4 --temp-dir /localhost/data/cassandra/tmp/ --verify --in-place
[2020-02-06 06:40:57,253] WARNING: is ccm : 0
[2020-02-06 06:40:57,283] INFO: Downloading data from backup to /localhost/data/cassandra/tmp/medusa-restore-efc46e67-f299-4e10-bc98-a3397c7fcf97
[2020-02-06 06:41:14,654] INFO: Stopping Cassandra
[2020-02-06 06:41:14,698] INFO: Moving backup data to Cassandra data directory
[2020-02-06 06:41:16,010] INFO: No --seeds specified so we will not wait for any
[2020-02-06 06:41:16,010] INFO: Starting Cassandra
[2020-02-06 06:41:16,024] INFO: Verifying the restore
[2020-02-06 06:41:16,024] INFO: Waiting for Cassandra to come up on localhost.localhost.net
[2020-02-06 06:41:17,833] INFO: Cassandra is up on localhost.localhost.net
[2020-02-06 06:41:17,834] INFO: Executing restore verify query: select * from tutorialspoint.emp;
Exception: Could not establish CQL session after 5

######## Error LOG

INFO  [StorageServiceShutdownHook] 2020-02-06 06:41:14,674 HintsService.java:209 - Paused hints dispatch
INFO  [StorageServiceShutdownHook] 2020-02-06 06:41:14,677 Server.java:179 - Stop listening for CQL clients
INFO  [StorageServiceShutdownHook] 2020-02-06 06:41:14,678 Gossiper.java:1647 - Announcing shutdown
INFO  [StorageServiceShutdownHook] 2020-02-06 06:41:14,679 StorageService.java:2442 - Node /127.0.0.1 state jump to shutdown
INFO  [StorageServiceShutdownHook] 2020-02-06 06:41:16,681 MessagingService.java:985 - Waiting for messaging service to quiesce
INFO  [ACCEPT-/127.0.0.1] 2020-02-06 06:41:16,682 MessagingService.java:1346 - MessagingService has terminated the accept() thread
INFO  [StorageServiceShutdownHook] 2020-02-06 06:41:16,988 HintsService.java:209 - Paused hints dispatch
ERROR [COMMIT-LOG-ALLOCATOR] 2020-02-06 06:41:16,993 StorageService.java:465 - Stopping gossiper
WARN  [COMMIT-LOG-ALLOCATOR] 2020-02-06 06:41:16,994 StorageService.java:322 - Stopping gossip by operator request
INFO  [COMMIT-LOG-ALLOCATOR] 2020-02-06 06:41:16,994 Gossiper.java:1647 - Announcing shutdown
INFO  [COMMIT-LOG-ALLOCATOR] 2020-02-06 06:41:16,995 StorageService.java:2442 - Node /127.0.0.1 state jump to shutdown
ERROR [StorageServiceShutdownHook] 2020-02-06 06:41:16,999 AbstractCommitLogSegmentManager.java:313 - Failed waiting for a forced recycle of in-use commit log segments
java.lang.AssertionError: attempted to delete non-existing file CommitLog-6-1580971059608.log
        at org.apache.cassandra.io.util.FileUtils.deleteWithConfirm(FileUtils.java:133) ~[apache-cassandra-3.11.5.jar:3.11.5]
        at org.apache.cassandra.io.util.FileUtils.deleteWithConfirm(FileUtils.java:160) ~[apache-cassandra-3.11.5.jar:3.11.5]
        at org.apache.cassandra.db.commitlog.CommitLogSegmentManagerStandard.discard(CommitLogSegmentManagerStandard.java:37) ~[apache-cassandra-3.11.5.jar:3.11.5]
        at org.apache.cassandra.db.commitlog.AbstractCommitLogSegmentManager.archiveAndDiscard(AbstractCommitLogSegmentManager.java:329) ~[apache-cassandra-3.11.5.jar:3.11.5]
        at org.apache.cassandra.db.commitlog.AbstractCommitLogSegmentManager.forceRecycleAll(AbstractCommitLogSegmentManager.java:303) ~[apache-cassandra-3.11.5.jar:3.11.5]
        at org.apache.cassandra.db.commitlog.CommitLog.forceRecycleAllSegments(CommitLog.java:208) [apache-cassandra-3.11.5.jar:3.11.5]
        at org.apache.cassandra.service.StorageService.drain(StorageService.java:4693) [apache-cassandra-3.11.5.jar:3.11.5]
        at org.apache.cassandra.service.StorageService$1.runMayThrow(StorageService.java:681) [apache-cassandra-3.11.5.jar:3.11.5]
        at org.apache.cassandra.utils.WrappedRunnable.run(WrappedRunnable.java:28) [apache-cassandra-3.11.5.jar:3.11.5]
        at org.apache.cassandra.concurrent.NamedThreadFactory.lambda$threadLocalDeallocator$0(NamedThreadFactory.java:84) [apache-cassandra-3.11.5.jar:3.11.5]
        at java.lang.Thread.run(Unknown Source) ~[na:1.8.0_171]
ERROR [COMMIT-LOG-ALLOCATOR] 2020-02-06 06:41:18,997 CommitLog.java:464 - Failed managing commit log segments. Commit disk failure policy is stop; terminating thread
org.apache.cassandra.io.FSWriteError: java.nio.file.NoSuchFileException: /localhost/data/cassandra/commitlog/CommitLog-6-1580971059610.log
        at org.apache.cassandra.db.commitlog.CommitLogSegment.<init>(CommitLogSegment.java:174) ~[apache-cassandra-3.11.5.jar:3.11.5]
        at org.apache.cassandra.db.commitlog.MemoryMappedSegment.<init>(MemoryMappedSegment.java:45) ~[apache-cassandra-3.11.5.jar:3.11.5]
        at org.apache.cassandra.db.commitlog.CommitLogSegment.createSegment(CommitLogSegment.java:131) ~[apache-cassandra-3.11.5.jar:3.11.5]
        at org.apache.cassandra.db.commitlog.CommitLogSegmentManagerStandard.createSegment(CommitLogSegmentManagerStandard.java:78) ~[apache-cassandra-3.11.5.jar:3.11.5]
        at org.apache.cassandra.db.commitlog.AbstractCommitLogSegmentManager$1.runMayThrow(AbstractCommitLogSegmentManager.java:110) ~[apache-cassandra-3.11.5.jar:3.11.5]
        at org.apache.cassandra.utils.WrappedRunnable.run(WrappedRunnable.java:28) [apache-cassandra-3.11.5.jar:3.11.5]
        at org.apache.cassandra.concurrent.NamedThreadFactory.lambda$threadLocalDeallocator$0(NamedThreadFactory.java:84) [apache-cassandra-3.11.5.jar:3.11.5]
        at java.lang.Thread.run(Unknown Source) ~[na:1.8.0_171]
Caused by: java.nio.file.NoSuchFileException: /localhost/data/cassandra/commitlog/CommitLog-6-1580971059610.log
        at sun.nio.fs.UnixException.translateToIOException(Unknown Source) ~[na:1.8.0_171]
        at sun.nio.fs.UnixException.rethrowAsIOException(Unknown Source) ~[na:1.8.0_171]
        at sun.nio.fs.UnixException.rethrowAsIOException(Unknown Source) ~[na:1.8.0_171]
        at sun.nio.fs.UnixFileSystemProvider.newFileChannel(Unknown Source) ~[na:1.8.0_171]
        at java.nio.channels.FileChannel.open(Unknown Source) ~[na:1.8.0_171]
        at java.nio.channels.FileChannel.open(Unknown Source) ~[na:1.8.0_171]
        at org.apache.cassandra.db.commitlog.CommitLogSegment.<init>(CommitLogSegment.java:169) ~[apache-cassandra-3.11.5.jar:3.11.5]
        ... 7 common frames omitted

--- a/medusa/cassandra_utils.py
+++ b/medusa/cassandra_utils.py
@@ -342,6 +342,10 @@ class Cassandra(object):
         return self._saved_caches_path
 
     @property
+    def hostname(self):
+        return self._hostname
+
+    @property
     def storage_port(self):
         return self._storage_port
 
@@ -533,28 +537,44 @@ class Cassandra(object):
 
 
 def wait_for_node_to_come_up(config, host, retries=10, delay=6):
-    """
-        Polls the node until the health check passes.
+    logging.info('Waiting for Cassandra to come up on %s', host)
+    check_function = is_node_up
+    if not _wait_for_node(config, host, check_function, retries, delay):
+        raise CassandraNodeNotUpError(host, retries)
+    logging.info('Cassandra is up on %s', host)
 
-        :param health_check: The type of health check to perform, one of cql, thrift, all.
+
+def wait_for_node_to_go_down(config, host, retries=10, delay=6):
+    logging.info('Waiting for Cassandra to go down on %s', host)
+
+    def check_function(config, host):
+        return not is_node_up(config, host)
+
+    if not _wait_for_node(config, host, check_function, retries, delay):
+        # TODO can do a kill here
+        raise CassandraNodeNotDownError(host, retries)
+    logging.info('Cassandra is down %s', host)
+
+
+def _wait_for_node(config, host, check_function, retries, delay):
+    """
+        Polls the node until the check_function check passes.
+
         :param host: The target host on which to run the check
+        :param check_function: called to determine node state
         :param retries: The number of times to retry the health check. Defaults to 10
         :param delay: A delay in seconds to wait before polling again. Defaults to 6 seconds.
-        :return: None when the node is determined to be up. If the retries are exhausted, an exception is raised.
-        """
-
-    logging.info('Waiting for Cassandra to come up on %s', host)
-
+        :return: True when the node is determined to be up. If the retries are exhausted, returns False
+    """
     attempts = 0
     while attempts < retries:
-        if is_node_up(config, host):
-            logging.info('Cassandra is up on %s', host)
-            return None
+        if check_function(config, host):
+            return True
         else:
             time.sleep(delay)
             attempts = attempts + 1
 
-    raise CassandraNodeNotUpError(host, attempts)
+    return False
 
 
 def is_node_up(config, host):
@@ -570,7 +590,7 @@ def is_node_up(config, host):
     must be ready to accept requests for both in order for the health check to be successful.
     """
 
-    health_check = config.restore.health_check
+    health_check = config.checks.health_check
     if int(config.cassandra.is_ccm) == 1:
         args = ['ccm', 'node1', 'nodetool']
         if health_check == 'thrift':
@@ -589,6 +609,7 @@ def is_node_up(config, host):
         elif health_check == 'all':
             return is_cassandra_up(list(args), rpc_port) and is_cassandra_up(list(args), native_port)
         else:
+            # cql only
             return is_cassandra_up(args, native_port)
 
 
@@ -615,7 +636,7 @@ def is_cassandra_up(args, port):
 
 class CassandraNodeNotUpError(Exception):
     """
-    Raised when it cannot be veriffied that a node is up by checking either nodetool statusbinary and/or
+    Raised when it cannot be verified that a node is up by checking either nodetool statusbinary and/or
     nodetool statusthrift
 
     Attributes:
@@ -626,3 +647,13 @@ class CassandraNodeNotUpError(Exception):
     def __init(self, host, attempts):
         msg = 'Could not verify that Cassandra is up on {host} after {attempts}'.format(host=host, attempts=attempts)
         super(CassandraNodeNotUpError, self).__init__(msg)
+
+
+class CassandraNodeNotDownError(Exception):
+    """
+    Raised when we give up waiting on a node to go down, meaning nodetool statusbinary and/or statusthrift keep
+    reporting open ports.
+    """
+    def __init(self, host, attempts):
+        msg = 'Could not verify that Cassandra is down on {host} after {attempts}'.format(host=host, attempts=attempts)
+        super(CassandraNodeNotDownError, self).__init__(msg)

--- a/medusa/config.py
+++ b/medusa/config.py
@@ -56,7 +56,7 @@ MonitoringConfig = collections.namedtuple(
 
 MedusaConfig = collections.namedtuple(
     'MedusaConfig',
-    ['storage', 'cassandra', 'ssh', 'restore', 'monitoring', 'logging']
+    ['storage', 'cassandra', 'ssh', 'checks', 'monitoring', 'logging']
 )
 
 LoggingConfig = collections.namedtuple(

--- a/medusa/restore_node.py
+++ b/medusa/restore_node.py
@@ -24,7 +24,7 @@ import sys
 import time
 import uuid
 
-from medusa.cassandra_utils import Cassandra, is_node_up
+from medusa.cassandra_utils import Cassandra, is_node_up, wait_for_node_to_go_down
 from medusa.download import download_data
 from medusa.storage import Storage
 from medusa.verify_restore import verify_restore
@@ -82,6 +82,7 @@ def restore_node_locally(config, temp_dir, backup_name, in_place, keep_auth, see
 
     logging.info('Stopping Cassandra')
     cassandra.shutdown()
+    wait_for_node_to_go_down(config, cassandra.hostname)
 
     # Clean the commitlogs, the saved cache to prevent any kind of conflict
     # especially around system tables.

--- a/medusa/verify_restore.py
+++ b/medusa/verify_restore.py
@@ -35,7 +35,7 @@ def verify_restore(hosts, config):
     for host in hosts:
         wait_for_node_to_come_up(config, host)
 
-    restore_verify_query = config.restore.query
+    restore_verify_query = config.checks.query
 
     if len(restore_verify_query) > 0:
         logging.info('Executing restore verify query: {}'.format(restore_verify_query))
@@ -47,9 +47,9 @@ def verify_restore(hosts, config):
             logging.info('Restore verify query completed successfully')
 
             # if config tells us to check for expected row count, we check that
-            expected_row_count = int(config.restore.expected_rows)
+            expected_row_count = int(config.checks.expected_rows)
             if expected_row_count:
-                if actual_row_count == int(config.restore.expected_rows):
+                if actual_row_count == int(config.checks.expected_rows):
                     logging.info('Restore verify query returned expected number of rows {}'.format(expected_row_count))
                 else:
                     logging.error('Restore verify query returned unexpected number of rows')
@@ -57,9 +57,9 @@ def verify_restore(hosts, config):
                     sys.exit(1)
 
             # if the config tells us to check for actual results, we check that too
-            if config.restore.expected_result:
+            if config.checks.expected_result:
                 # expected result allows just one row
-                expected_result = config.restore.expected_result
+                expected_result = config.checks.expected_result
                 actual_result = ','.join(rows[0] if len(rows) >= 1 else [])
                 if actual_result == expected_result:
                     logging.info('Restore verify query returned expected result {}'.format(expected_result))

--- a/tests/cassandra_utils_test.py
+++ b/tests/cassandra_utils_test.py
@@ -42,7 +42,7 @@ class CassandraUtilsTest(unittest.TestCase):
             monitoring={},
             cassandra=None,
             ssh=None,
-            restore=None,
+            checks=None,
             logging=None
         )
 
@@ -134,7 +134,7 @@ class CassandraUtilsTest(unittest.TestCase):
             monitoring=None,
             cassandra=_namedtuple_from_dict(CassandraConfig, config['cassandra']),
             ssh=None,
-            restore=None,
+            checks=None,
             logging=None
         )
         n = Nodetool(medusa_config.cassandra).nodetool
@@ -155,7 +155,7 @@ class CassandraUtilsTest(unittest.TestCase):
             monitoring=None,
             cassandra=_namedtuple_from_dict(CassandraConfig, config['cassandra']),
             ssh=None,
-            restore=None,
+            checks=None,
             logging=None
         )
         n = Nodetool(medusa_config.cassandra).nodetool
@@ -178,7 +178,7 @@ class CassandraUtilsTest(unittest.TestCase):
             monitoring=None,
             cassandra=_namedtuple_from_dict(CassandraConfig, config['cassandra']),
             ssh=None,
-            restore=None,
+            checks=None,
             logging=None
         )
         n = Nodetool(medusa_config.cassandra).nodetool
@@ -203,7 +203,7 @@ class CassandraUtilsTest(unittest.TestCase):
             monitoring=None,
             cassandra=_namedtuple_from_dict(CassandraConfig, config['cassandra']),
             ssh=None,
-            restore=None,
+            checks=None,
             logging=None
         )
 
@@ -234,7 +234,7 @@ class CassandraUtilsTest(unittest.TestCase):
             monitoring=None,
             cassandra=_namedtuple_from_dict(CassandraConfig, config['cassandra']),
             ssh=None,
-            restore=None,
+            checks=None,
             logging=None
         )
 
@@ -266,7 +266,7 @@ class CassandraUtilsTest(unittest.TestCase):
             monitoring=None,
             cassandra=_namedtuple_from_dict(CassandraConfig, config['cassandra']),
             ssh=None,
-            restore=None,
+            checks=None,
             logging=None
         )
 
@@ -297,7 +297,7 @@ class CassandraUtilsTest(unittest.TestCase):
             monitoring=None,
             cassandra=_namedtuple_from_dict(CassandraConfig, config['cassandra']),
             ssh=None,
-            restore=None,
+            checks=None,
             logging=None
         )
 

--- a/tests/integration/features/steps/integration_steps.py
+++ b/tests/integration/features/steps/integration_steps.py
@@ -227,12 +227,16 @@ def i_am_using_storage_provider(context, storage_provider, client_encryption):
 
     config["monitoring"] = {"monitoring_provider": "local"}
 
+    config["checks"] = {
+        "health_check": "cql"
+    }
+
     context.medusa_config = MedusaConfig(
         storage=_namedtuple_from_dict(StorageConfig, config["storage"]),
         cassandra=_namedtuple_from_dict(CassandraConfig, config["cassandra"]),
         monitoring=_namedtuple_from_dict(MonitoringConfig, config["monitoring"]),
         ssh=None,
-        restore=None,
+        checks=_namedtuple_from_dict(ChecksConfig, config["checks"]),
         logging=None
     )
     cleanup_storage(context, storage_provider)
@@ -736,7 +740,7 @@ def _i_can_verify_the_restore_verify_query_returned_rows(context, query, expecte
         storage=context.medusa_config.storage,
         cassandra=context.medusa_config.cassandra,
         monitoring=context.medusa_config.monitoring,
-        restore=_namedtuple_from_dict(ChecksConfig, restore_config),
+        checks=_namedtuple_from_dict(ChecksConfig, restore_config),
         ssh=None,
         logging=None
     )

--- a/tests/purge_test.py
+++ b/tests/purge_test.py
@@ -44,7 +44,7 @@ class PurgeTest(unittest.TestCase):
             monitoring={},
             cassandra=None,
             ssh=None,
-            restore=None,
+            checks=None,
             logging=None,
         )
         self.storage = Storage(config=self.config.storage)

--- a/tests/restore_cluster_test.py
+++ b/tests/restore_cluster_test.py
@@ -48,7 +48,7 @@ class RestoreClusterTest(unittest.TestCase):
             monitoring={},
             cassandra=_namedtuple_from_dict(CassandraConfig, config['cassandra']),
             ssh=None,
-            restore=None,
+            checks=None,
             logging=None
         )
 

--- a/tests/restore_node_test.py
+++ b/tests/restore_node_test.py
@@ -36,7 +36,7 @@ class RestoreNodeTest(unittest.TestCase):
             monitoring={},
             cassandra=None,
             ssh=None,
-            restore=None,
+            checks=None,
             logging=None
         )
 

--- a/tests/storage_test.py
+++ b/tests/storage_test.py
@@ -63,7 +63,7 @@ class RestoreNodeTest(unittest.TestCase):
             cassandra=_namedtuple_from_dict(CassandraConfig, config['cassandra']),
             monitoring={},
             ssh=None,
-            restore=None,
+            checks=None,
             logging=None
         )
 

--- a/tests/storage_test_with_prefix.py
+++ b/tests/storage_test_with_prefix.py
@@ -64,7 +64,7 @@ class RestoreNodeTest(unittest.TestCase):
             cassandra=_namedtuple_from_dict(CassandraConfig, config['cassandra']),
             monitoring={},
             ssh=None,
-            restore=None,
+            checks=None,
             logging=None
         )
 

